### PR TITLE
Fix tree view reveal() behavior

### PR DESCRIPTION
**Pull request by amiramw**
Signed-off-by: Amiram Wingarten <amiram.wingarten@sap.com>

#### What it does

Various fixes to support tree view reveal in several unsupported scenarios and align it with vscode.

- reveal called without focus now opens the view container (left pane) and view (section), without focus
- reveal of a node that its ancestors are collapsed now expand all parents and reveals it
- reveal of a node that was not cached by backend or not fetched by client (for example when showing the tree for the first time) now works
- reveal of a tree element that is not identical (`===`) to cached node now works

#### How to test

- Clone https://github.com/amiramw/vscode-tree-expand
- `yarn && vsce package` to create the vsix file
- Load the vsix in theia
- Test view tree will show in explorer view a tree with nodes **a ➡️ b ➡️ c**

##### Scenarios

First time:
1. Open theia from chrome incognito (like first time).
2. F1 -> Reveal c New Object or F1 -> Reveal c Same Object
3. Test view section is opened, a and b are expanded and c is selected, before this PR nothing happened

From other view container:
1. Go to debug pane.
2. F1 -> Reveal c New Object or F1 -> Reveal c Same Object
3. Explorer is opened. Test view section is opened, a and b are expanded and c is selected, before this PR nothing happened.

From other view container:
1. Go to debug pane.
2. F1 -> Reveal c New Object or F1 -> Reveal c Same Object
3. Explorer is opened. Test view section is opened, a and b are expanded and c is selected, before this PR nothing happened.

With non identical element:
1. Open test view 
2. Expand the tree to show c and then collapse it
3. F1 -> Reveal c New Object or F1 -> Reveal c Same Object
4. Tree is expanded and c is selected. Before this PR only "Reveal c Same Object" worked

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by amiramw** [comment]
@kittaakos @vince-fugnitto do one of you has the time to review this? 🙏

**Comment by kittaakos** [comment]
@amiramw, I'll look into the changes today. I tried it yesterday and it worked. 👍 Regarding the API; instead of the paren-chain string array we pass around, can we support the same API as VS Code does?
```ts
$reveal(treeViewId: string, itemInfo: { item: ITreeItem, parentChain: ITreeItem[] } | undefined, options: IRevealOptions): Promise<void>;
```
Do you know why cannot we have a reference to the `treeDataProvider` in `plugin-view-registry.ts`? Calculating the parent chain in place would be much better. What do you think?

**Comment by amiramw** [comment]
> Regarding the API; instead of the paren-chain string array we pass around, can we support the same API as VS Code does?
> 
> ```ts
> $reveal(treeViewId: string, itemInfo: { item: ITreeItem, parentChain: ITreeItem[] } | undefined, options: IRevealOptions): Promise<void>;
> ```

I don't think our tree handling is exact copy of vscode. I did see the parent chain and thought that we need it too. But currently there are a lot of fields in `ITreeItem` that are not relevant to the interface.

> Do you know why cannot we have a reference to the `treeDataProvider` in `plugin-view-registry.ts`? Calculating the parent chain in place would be much better. What do you think?

Because `treeDataProvider` is defined in the plugin host process and `plugin-view-registry.ts` is on the browser. So there can't be in place calls to it.

File: packages/plugin-ext/src/plugin/tree/tree-views.ts [hunk 1]
```diff
@@ -199,15 +199,9 @@ class TreeViewExtImpl<T> implements Disposable {
...
-        this.nodes.forEach((el, id) => {
-            if (Object.is(el.value, element)) {
-                elementId = id;
-            }
-        });
-
-        if (elementId) {
-            return this.proxy.$reveal(this.treeViewId, elementId, {
+        const elementParentChain = await this.calculateRevealParentChain(element);
+        if (elementParentChain) {
```
Comment (kittaakos) [inline packages/plugin-ext/src/plugin/tree/tree-views.ts:203]: What should be revealed when the `elementParentChain` is `['']`?

File: packages/plugin-ext/src/main/browser/view/tree-views-main.ts [hunk 2]
```diff
@@ -98,14 +103,17 @@ export class TreeViewsMainImpl implements TreeViewsMain, Disposable {
     }
 
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    async $reveal(treeViewId: string, treeItemId: string, options: TreeViewRevealOptions): Promise<any> {
-        const viewPanel = await this.viewRegistry.openView(treeViewId, { activate: options.focus });
+    async $reveal(treeViewId: string, elementParentChain: string[], options: TreeViewRevealOptions): Promise<any> {
```
Comment (kittaakos) [inline packages/plugin-ext/src/main/browser/view/tree-views-main.ts:106]: Please add a single line of comment on what `elementParentChain` is. The original `treeItemId` was obvious. Thanks!

**Comment by kittaakos** [comment]
@amiramw, I added a few remarks. Please adjust. The behavior works, but I will verify once more.

I do not know if we have to note the breaking API change in the changelog, but doing it should not hurt. Please do so. Thanks!

File: packages/plugin-ext/src/plugin/tree/tree-views.ts [hunk 3]
```diff
@@ -238,6 +232,77 @@ class TreeViewExtImpl<T> implements Disposable {
...
+        if (!element) {
+            // root
+            return [''];
+        }
+        const parent = this.treeDataProvider.getParent && await this.treeDataProvider.getParent(element);
+        const chain = await this.calculateRevealParentChain(parent);
+        if (!chain || chain.length === 0) {
+            // parents are inconsistent
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (amiramw) [reply packages/plugin-ext/src/plugin/tree/tree-views.ts:254]: Changed now so technical root doesn't get an entry in the array

File: packages/plugin-ext/src/plugin/tree/tree-views.ts [hunk 4]
```diff
@@ -238,6 +232,77 @@ class TreeViewExtImpl<T> implements Disposable {
...
+        let possibleIndex = children.length;
+        // find the right element id by searching all possible id names in the cache
+        while (possibleIndex-- > 0) {
+            const candidateId = this.buildTreeItemId(parentId, possibleIndex, idLabel);
+            if (this.nodes.has(candidateId)) {
+                return chain.concat(candidateId);
+            }
+        }
+        // couldn't calculate consistent parent chain and id
+        return;
```
[the remark this answers is not in the document]
Comment (amiramw) [reply packages/plugin-ext/src/plugin/tree/tree-views.ts:275]: done

**Comment by amiramw** [comment]
> I do not know if we have to note the breaking API change in the changelog, but doing it should not hurt. Please do so. Thanks!

@kittaakos I fixed all the code comments. Which API do you mean?

**Comment by kittaakos** [comment]
> I fixed all the code comments.

Thank you! 👍 

> Which API do you mean?

I meant this one. The second arg was a string.
```
async $reveal(treeViewId: string, elementParentChain: string[], options: TreeViewRevealOptions): Promise<any>
```

I leave the decision to you.

**Comment by amiramw** [comment]
@kittaakos I added the changelog as well

**Review by kittaakos — APPROVED** [review]
Please proceed with the merge, @amiramw 👍 

(Two explicit `return undefined` is missing, though.)

**Comment by amiramw** [comment]
> Please proceed with the merge, @amiramw 👍
> 
> (Two explicit `return undefined` is missing, though.)

Fixed. I'll merge once builds finish.

File: packages/plugin-ext/src/main/browser/view/tree-views-main.ts (hunk 2 above)
Comment (tsmaeder) [reply packages/plugin-ext/src/main/browser/view/tree-views-main.ts:108]: May I suggest a rename to "ancestorNodeIds"

Comment (amiramw) [reply packages/plugin-ext/src/main/browser/view/tree-views-main.ts:108]: the term parent chain is also used in code source: https://github.com/microsoft/vscode/blob/1aa76d792f6857a0444b71e0a169f34a8ef5488d/src/vs/workbench/api/common/extHostTreeViews.ts#L390

**Pull request merged**